dbms/src/Common/Spiller.cpp

```diff
@@ -0,0 +1,181 @@
+{}
+
+SpilledFile::~SpilledFile()
+{
+    try
+    {
+        if (exists())
+            remove(true);
+    }
+    catch (...)
+    {
+    }
+}
+
+SpilledFilesInputStream::SpilledFilesInputStream(const std::vector<String> & spilled_files_, const Block & header_)
+    : spilled_files(spilled_files_)
+    , header(header_)
+{
+    RUNTIME_CHECK_MSG(!spilled_files.empty(), "Spilled files must not be empty");
+    current_reading_file_index = 0;
+    current_file_stream = std::make_unique<SpilledFileStream>(spilled_files[0], header);
+}
+
+Block SpilledFilesInputStream::readImpl()
+{
+    Block ret;
+    ret = current_file_stream->block_in->read();
+    if (ret)
+        return ret;
+    current_reading_file_index++;
+    for (; current_reading_file_index < spilled_files.size(); current_reading_file_index++)
+    {
+        current_file_stream = std::make_unique<SpilledFileStream>(spilled_files[current_reading_file_index], header);
+        ret = current_file_stream->block_in->read();
+        if (ret)
+            return ret;
+    }
+    return ret;
+}
+
+Block SpilledFilesInputStream::getHeader() const
+{
+    return header;
+}
+String SpilledFilesInputStream::getName() const
+{
+    return "SpilledFiles";
+}
+
+Spiller::Spiller(const String & id_, bool is_input_sorted_, size_t partition_num_, const String & spill_dir_, const Block & input_schema_, LoggerPtr logger_)
+    : id(id_)
+    , is_input_sorted(is_input_sorted_)
+    , partition_num(partition_num_)
+    , spill_dir(spill_dir_)
+    , input_schema(input_schema_)
+    , logger(logger_)
+{
+    if (spill_dir.at(spill_dir.size() - 1) != Poco::Path::separator())
+    {
+        spill_dir += Poco::Path::separator();
+    }
+    spilled_files.resize(partition_num);
+}
+
+bool Spiller::spillBlocks(const Blocks & blocks, size_t partition_id)
+{
+    RUNTIME_CHECK_MSG(partition_id < partition_num, "{}: partition id {} exceeds partition num {}.", id, partition_id, partition_num);
+    RUNTIME_CHECK_MSG(spill_finished == false, "{}: spill after the spiller is finished.", id);
+    /// todo append to existing file
+    auto spilled_file_name = nextSpillFileName(partition_id);
+    try
+    {
+        auto spilled_file = std::make_unique<SpilledFile>(spilled_file_name);
+        if (spilled_file->exists())
+            throw Exception("Duplicated spilled files, should not happens");
+        WriteBufferFromFile file_buf(spilled_file_name);
+        CompressedWriteBuffer compressed_buf(file_buf);
+        NativeBlockOutputStream block_out(compressed_buf, 0, blocks[0].cloneEmpty());
+        block_out.writePrefix();
+        for (const auto & block : blocks)
+        {
+            auto block_bytes_size = block.bytes();
+            block_out.write(block);
+            spilled_file->addSpilledDataSize(block_bytes_size);
+        }
+        {
+            std::lock_guard lock(spilled_files_mutex);
+            spilled_files[partition_id].emplace_back(std::move(spilled_file));
```

Comment:
When do we delete the temporary files?

Reply:
The temporary files will be deleted in the destructor of `SpilledFile`